src/test/java/org/prebid/server/ApplicationTest.java

```diff
@@ -246,193 +489,543 @@ public void openrtb2AuctionShouldRespondWithBidsFromDifferentExchanges() throws
+
+        //then
+        final String expectedAuctionResponse = auctionResponseFrom(jsonFrom("openrtb2/beachfront/test-auction-beachfront-response.json"),
+                response,  "ext.responsetimemillis.%s", exchanges);
+
+        JSONAssert.assertEquals(expectedAuctionResponse, response.asString(), JSONCompareMode.NON_EXTENSIBLE);
+    }
 
+    @Test
+    public void openrtb2AuctionShouldRespondWithBidsFromAdkerneladn() throws IOException, JSONException {
+        // given
         // adkernelAdn bid response for imp 021
         wireMockRule.stubFor(post(urlPathEqualTo("/adkerneladn-exchange"))
                 .withQueryParam("account", equalTo("101"))
                 .withHeader("Content-Type", equalToIgnoreCase("application/json;charset=UTF-8"))
                 .withHeader("Accept", equalTo("application/json"))
                 .withHeader("x-openrtb-version", equalTo("2.5"))
-                .withRequestBody(equalToJson(jsonFrom("openrtb2/test-adkerneladn-bid-request-1.json")))
-                .willReturn(aResponse().withBody(jsonFrom("openrtb2/test-adkerneladn-bid-response-1.json"))));
+                .withRequestBody(equalToJson(jsonFrom("openrtb2/adkerneladn/test-adkerneladn-bid-request-1.json")))
+                .willReturn(aResponse().withBody(jsonFrom("openrtb2/adkerneladn/test-adkerneladn-bid-response-1.json"))));
 
         // adkernelAdn bid response for imp 022
         wireMockRule.stubFor(post(urlPathEqualTo("/adkerneladn-exchange"))
                 .withQueryParam("account", equalTo("102"))
                 .withHeader("Content-Type", equalToIgnoreCase("application/json;charset=UTF-8"))
                 .withHeader("Accept", equalTo("application/json"))
                 .withHeader("x-openrtb-version", equalTo("2.5"))
-                .withRequestBody(equalToJson(jsonFrom("openrtb2/test-adkerneladn-bid-request-2.json")))
-                .willReturn(aResponse().withBody(jsonFrom("openrtb2/test-adkerneladn-bid-response-2.json"))));
+                .withRequestBody(equalToJson(jsonFrom("openrtb2/adkerneladn/test-adkerneladn-bid-request-2.json")))
+                .willReturn(aResponse().withBody(jsonFrom("openrtb2/adkerneladn/test-adkerneladn-bid-response-2.json"))));
+
+        // pre-bid cache
+        wireMockRule.stubFor(post(urlPathEqualTo("/cache"))
+                .withRequestBody(equalToJson(jsonFrom("openrtb2/adkerneladn/test-cache-adkerneladn-request.json")))
+                .willReturn(aResponse().withBody(jsonFrom("openrtb2/adkerneladn/test-cache-adkerneladn-response.json"))));
+
+        final Map<String, String> exchanges = new HashMap<>();
+        exchanges.put(ADKERNELADN, "http://localhost:" + WIREMOCK_PORT + "/adkerneladn-exchange");
+
+        //when
+        final Response response = given(spec)
+                .header("Referer", "http://www.example.com")
+                .header("X-Forwarded-For", "192.168.244.1")
+                .header("User-Agent", "userAgent")
```

Comment:
There's no uids cookie - is this intentional?

Reply:
no

Reply:
Please address this comment